File: flatten.py
#Time complexity: O(n); n = no. of nested integers
#Space complexity: O(d); d = depth of largest nested integer
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.st = []
        self.nextEle = None
        self.st.append(iter(nestedList))
        
    
    def next(self) -> int:
        return self.nextEle.getInteger()
        
    
    def hasNext(self) -> bool:
        while self.st:
            self.nextEle = next(self.st[-1], None)
            if not self.nextEle:
                self.st.pop()
            elif self.nextEle.isInteger():
                return True
            else:
                self.st.append(iter(self.nextEle.getList()))
        return False
        
         

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: Flatten.java
//Time complexity-O(n)
//Space complexity-O(n)
//Ran on leetcode-yes
//Solution with code- 

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return empty list if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */
public class NestedIterator implements Iterator<Integer> {
    Queue<Integer> q ;//storing all value to the queue first
    
    public NestedIterator(List<NestedInteger> nestedList) {
            q= new LinkedList<>();
            dfs(nestedList);
    }
    
    public void dfs(List<NestedInteger> elements){
        for(NestedInteger n: elements){
            if(n.isInteger()){
                q.add(n.getInteger());
            }
            else{
                dfs(n.getList());//dfs for list in list
            }
        }
    }

    @Override
    public Integer next() {
        return q.poll();
    }

    @Override
    public boolean hasNext() {
        return !q.isEmpty();
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */